ocaml-lsp-server/src/hover_req.ml

```diff
@@ -12,31 +12,78 @@ let environment_mode =
   | Some true -> Extended_variable
   | Some false | None -> Default
 
-let format_contents ~syntax ~markdown ~typ ~doc =
+let format_contents ~syntax ~markdown ~typ ~doc ~syntax_doc_name ~syntax_doc_des
+    ~syntax_doc_url =
   (* TODO for vscode, we should just use the language id. But that will not work
      for all editors *)
   `MarkupContent
     (if markdown then
-       let value =
-         let markdown_name = Document.Syntax.markdown_name syntax in
-         match doc with
-         | None -> sprintf "```%s\n%s\n```" markdown_name typ
-         | Some s ->
-           let doc =
-             match Doc_to_md.translate s with
-             | Raw d -> sprintf "(** %s *)" d
-             | Markdown d -> d
-           in
-           sprintf "```%s\n%s\n```\n---\n%s" markdown_name typ doc
-       in
-       { MarkupContent.value; kind = MarkupKind.Markdown }
-     else
-       let value =
-         match doc with
-         | None -> sprintf "%s" typ
-         | Some d -> sprintf "%s\n%s" typ d
-       in
-       { MarkupContent.value; kind = MarkupKind.PlainText })
+     let syntax_doc_name =
+       match Doc_to_md.translate syntax_doc_name with
+       | Raw d -> sprintf "(** %s *)" d
+       | Markdown d -> d
+     in
+     let syntax_doc_des =
+       match Doc_to_md.translate syntax_doc_des with
+       | Raw d -> sprintf "(** %s *)" d
+       | Markdown d -> d
+     in
+     let syntax_doc_url =
+       match Doc_to_md.translate syntax_doc_url with
+       | Raw d -> sprintf "(** %s *)" d
+       | Markdown d -> d
+     in
+     let value =
+       let markdown_name = Document.Syntax.markdown_name syntax in
+       match doc with
+       | None ->
+         sprintf
+           "```%s\n%s\n```\n---\n%s %s\n\n---\n\n%s"
+           markdown_name
+           typ
+           (if String.length syntax_doc_name == 0 then ""
+           else syntax_doc_name ^ ":")
+           syntax_doc_des
+           (if String.length syntax_doc_url == 0 then ""
+           else "Go to [Documentation](" ^ syntax_doc_url ^ ")")
+       | Some s ->
+         let doc =
+           match Doc_to_md.translate s with
+           | Raw d -> sprintf "(** %s *)" d
+           | Markdown d -> d
+         in
+         sprintf
+           "```%s\n%s\n```\n---\n%s %s\n\n---\n\n%s\n\n---\n\n%s"
+           markdown_name
+           typ
+           (if String.length syntax_doc_name == 0 then ""
+           else syntax_doc_name ^ ":")
+           syntax_doc_des
+           (if String.length syntax_doc_url == 0 then ""
```

Comment:
You should not have to rely on such tests. Merlin commits to always give a manual page url so either there is no documentation at all, or the url should not be empty.